src/frontier_detection/src/map.cpp

```diff
@@ -0,0 +1,278 @@
+            auto it = globalMap->begin_leafs();
+            std::advance(it, index);
+
+            size_t processed = 0;
+            for (; it != globalMap->end_leafs() && processed < chunk_size; ++it, ++processed) {
+                // Validate occupancy value
+                if (std::isnan(it->getOccupancy())) {
+                    // Log the invalid node details
+                    std::lock_guard<std::mutex> lock(log_mutex);
+                    RCLCPP_WARN(
+                        logger,
+                        "Invalid node detected. Coordinates: [%f, %f, %f], Occupancy: NaN",
+                        it.getCoordinate().x(),
+                        it.getCoordinate().y(),
+                        it.getCoordinate().z()
+                    );
+
+                    // Remove the invalid node or reset its value
+                    globalMap->deleteNode(it.getKey());  // Remove the node
+                }
+            }
+            return true;
+        }));
+        index += chunk_size;
+    }
+
+    // Wait for all tasks and aggregate results
+    for (auto& task : tasks) {
+        if (!task.get()) {
+            return false; // If any thread fails validation
+        }
+    }
+
+    return true; // All nodes are valid
+}
+
+OctoMapNode::OctoMapNode(std::string nodeName): Node(nodeName){
+	RCLCPP_INFO(this->get_logger(), "Creating OctoMap Node");
+	mpGlobalMap = std::make_shared<OctoMap>(this->get_logger());
+    const int num_channels = 3; // x y z
+    mpRawPointCloud.header.frame_id = "map";
+    mpRawPointCloud.height = 1;
+    mpRawPointCloud.width = 0;
+    mpRawPointCloud.is_bigendian = false;
+    mpRawPointCloud.is_dense = true;
+    mpRawPointCloud.point_step = num_channels * sizeof(float);
+    mpRawPointCloud.row_step = mpRawPointCloud.point_step * mpRawPointCloud.width;
+    mpRawPointCloud.fields.resize(num_channels);
+    std::string channel_id[] = { "x", "y", "z"};
+    for (int i = 0; i<num_channels; i++) {
+        mpRawPointCloud.fields[i].name = channel_id[i];
+        mpRawPointCloud.fields[i].offset = i * sizeof(float);
+        mpRawPointCloud.fields[i].count = 1;
+        mpRawPointCloud.fields[i].datatype = sensor_msgs::msg::PointField::FLOAT32;
+    }
+
+    mpRawPointCloud.data.resize(mpRawPointCloud.row_step * mpRawPointCloud.height);
+	mpPointCloudSubscriber = this->create_subscription<MapMsg>(
+			"/orbslam3_mono_node/map_points",
+			10,
+			std::bind(&OctoMapNode::AddNewPointCloudCallback, this, std::placeholders::_1));
+	mpMapPublisher = this->create_publisher<octomap_msgs::msg::Octomap>("~/octomap", 10);
+	mpPointCloudPublisher = this->create_publisher<sensor_msgs::msg::PointCloud2>("~/point_cloud", 10);
+	mpMapPublisherTimer = this->create_wall_timer(
+		std::chrono::milliseconds(100), std::bind(&OctoMapNode::PublishMap, this)
+	);
+	mpThrMsgQueueProcess = new std::thread(&OctoMapNode::UpdateMap, this);	
```

Comment:
memory leak here! why do we need to assign a pointer, we can just create a std::thread object?

Reply:
I wanted to clean up this up in the destructor. let me check if i did I am calling join within the destructor.

Reply:
I did not clean up the thread there. Also now that I think about it could have been a std::thread object too here.